core/device_hooks/common_kernels.inc.cpp

```diff
@@ -136,6 +138,11 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
     _macro(ValueType, IndexType) GKO_NOT_COMPILED(GKO_HOOK_MODULE); \
     GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(_macro)
 
+#define GKO_STUB_VALUE_AND_INT32_TYPE(_macro)                       \
+    template <typename ValueType, typename IndexType>               \
+    _macro(ValueType, IndexType) GKO_NOT_COMPILED(GKO_HOOK_MODULE); \
+    GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INT32_TYPE(_macro)
```

Comment:
I think it is good to keep the public interface more general and instantiate later for int64 or other int variants if necessary.

Reply:
I think 64 bit indices is very unrealistic. 2^31 rows easily saturate an entire GPU (for the forseeable future), all the advantages of batching (fast local memory, fast synchronization, load balancing via oversubscription) don't seem to apply.

Reply:
But it might make sense to use int16, int8 etc. I think keeping the option open makes sense. We instantiate only for int32, so we dont have any downsides such as larger library objects etc, but the public interface remains extensible.

Reply:
I agree that interpreting batch objects as regular Ginkgo objects is useful. My point is that you have the option of different IndexType for the monolithic matrix format, but if you remove the IndexType option, then that requires a complete complete re-write of the class.

Reply:
Does this really require a full rewrite? I would expect only having to remove the `typename IndexType` template parameter and replacing all remaining instances of `IndexType` by `index_type` or `int32` to be very doable.

Reply:
I think if we add another IndexType in the future, we would add it to `matrix::Ell` as well.

Reply:
Which other index type do you have in mind? int16 doesn't make sense for the scales we are looking at in Ginkgo, int128 is way beyond even future system capabilities.

Reply:
`int16` definitely makes sense for batched atleast. To ensure interoperability with batched, we can instantiate non-batched for `int16` as well.

Reply:
Adding another index type is a huge increase in build time in Ginkgo, and int16 just doesn't make sense for the scales of sparse linear algebra we look at. If you want to use 16 bit indexing inside a solver kernel, a simple solution is to convert the indices on-the-fly in temporary memory - the sparsity pattern is tiny compared to the rest.